File: Problems/CombinationSum3/solution.cpp
class Solution {
public:

    std::vector<std::vector<int>> combinationSum3(int k, int n) {
        std::vector<int> candidates { 1, 2, 3, 4, 5, 6, 7, 8, 9 };
        std::vector<std::vector<int>> output;
        std::vector<int> startCombination;
        backtrack(candidates, n, k, startCombination, 0, output);
        return output;
    }

    void backtrack(std::vector<int>& candidates, int target, int limit, std::vector<int>& combination, int startIndex, std::vector<std::vector<int>>& output) {

        if (target == 0 && combination.size() == limit) {
            std::vector<int> combinationCopy(combination);
            output.push_back(combinationCopy);
            return;
        }

        if (target < 0) {
            return;
        }

        for (int i = startIndex; i < candidates.size(); ++i) {

            if (i > startIndex && candidates[i] == candidates[i - 1]) {
                continue;
            }

            combination.push_back(candidates[i]);
            backtrack(candidates, target - candidates[i], limit, combination, i + 1, output);
            combination.pop_back();
        }

    }
};

File: Problems/CombinationSum3/solution.py
class Solution(object):

    def combination_sum_3(self, k, n):
        # Define the candidate numbers from 1 to 9
        CANDIDATES = [1, 2, 3, 4, 5, 6, 7, 8, 9]

        output = []  # This will store all valid combinations

        # Start backtracking from index 0 with an empty combination
        self.backtrack(0, n, k, CANDIDATES, output, [])

        return output
    
    def backtrack(self, start_index, target, limit, candidates, output, combination):
        # Base case: valid combination found
        if target == 0 and len(combination) == limit:
            output.append(list(combination))  # Add a copy of the current combination
            return

        # If the remaining target is negative, prune this path
        if target < 0:
            return
        
        # Explore all candidates starting from current index
        for i in range(start_index, len(candidates)):

            # Optional: skip duplicates (not strictly needed since candidates are unique)
            if i > start_index and candidates[i] == candidates[i - 1]:
                continue

            # Choose the current number
            combination.append(candidates[i])

            # Recurse: move to the next index, reduce the target
            self.backtrack(i + 1, target - candidates[i], limit, candidates, output, combination)

            # Backtrack: remove the last number to try other paths
            combination.pop()

solution = Solution()

assert(solution.combination_sum_3(3, 7) == [[1,2,4]])
assert(solution.combination_sum_3(3, 9) == [[1,2,6],[1,3,5],[2,3,4]])

print("PASS")